p4_symbolic/symbolic/parser.cc

```diff
@@ -0,0 +1,165 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Hardcodes the behavior of an interesting p4 parser that is part
+// of the p4 program we are interested in.
+// The hardcoded behavior sets the validity of certain header fields
+// based on the fields in the packet, and sets the default values
+// for local_metadata fields.
+
+#include "p4_symbolic/symbolic/parser.h"
+
+#include "z3++.h"
+
+namespace p4_symbolic {
+namespace symbolic {
+namespace parser {
+
+gutil::StatusOr<std::vector<z3::expr>> EvaluateHardcodedParser(
+    SymbolicPerPacketState *state) {
+  std::vector<z3::expr> constraints;
+  // Set initial values for certain special metadata fields.
+  z3::expr true_guard = Z3Context().bool_val(true);
+  if (state->ContainsKey("local_metadata.vrf_id")) {
+    RETURN_IF_ERROR(state->Set("vrf_id", Z3Context().bv_val(0, 1), true_guard));
+  }
+  if (state->ContainsKey("local_metadata.l4_src_port")) {
+    RETURN_IF_ERROR(state->Set("local_metadata.l4_src_port",
+                               Z3Context().bv_val(0, 1), true_guard));
+  }
+  if (state->ContainsKey("local_metadata.l4_dst_port")) {
```

Comment:
It would be nice if we returned an error  if one of the variable names doesn't exist, instead of just failing silently. I understand that maybe this will break some of your  tests that use simpler P4 programs. Should we maybe only do the parser constraints for the real program? Maybe we could indeed put them into the assertions? Maybe we should pass a flag to the solver of whether to make the assertions or not?

Reply:
This is a good question.
If we make this fail it will break my tests. Passing in a flag might be a good idea here (like a prod vs dev flag or something).

I disagree with the phrasing of "failing silently", if these variables do not exist, it means that the program being analyzed does not have them or use them, we can still evaluate it, so nothing has failed up to that point. It is just that we do not support automatic handling of parsers (which is actually easy to do), and users should expect that for some arbitrary parser, we cannot capture its behavior.

I am unsure what to do. The passing a flag idea might be ok, since I do not expect we will use this tool in production for programs with a parser that is different than the hard-coded one. So I guess that would probably be best?